ml4h/evaluate_segmentations.py

```diff
@@ -0,0 +1,141 @@
+        data_generator: list, 
+        metrics: List[Callable] = None):
+    # Reformat metrics list to a metrics dictionary of the form
+    # {name: function pointer}.
+    if metrics is not None:
+        if isinstance(metrics, list):
+            metrics_dict = {m.name: m for m in metrics}
+        else:
+            metrics_dict = metrics
+    else:
+        metrics_dict = {m.name: m for m in model.metrics}
+    # Reload model
+    model.compile(optimizer=model.optimizer,
+                        loss=model.loss,
+                        metrics=list(metrics_dict.values()))
+    #
+    # Logger for outputs
+    logger = BatchMetricsLogger(metrics = metrics_dict)
+    try:
+        eval = model.evaluate(data_generator, callbacks=[logger])
+    except Exception as e:
+        raise Exception(f'Failed to evaluate model:\n{e}')
+    eval_batch = pd.DataFrame(logger.storage, index = np.arange(len(logger.storage)))
+    # Workaround for old Keras
+    eval = {out: eval[i] for i, out in enumerate(model.metrics_names)}
+    eval = pd.DataFrame(eval,index=[0])
+    # Make predictions
+    prediction_prob = model.predict(data_generator)
+    # Assumes channel_order
+    predictions = tf.argmax(prediction_prob, -1)
+    # Return eval, eval_batch pair
+    return eval, eval_batch, prediction_prob, predictions
+
+
+def evaluate_segmentation_models(models: List[tf.keras.models.Model], 
+        data_generators: list, 
+        metrics: List[Callable] = None,
+        channel_order: str = 'channels_last'): # Todo: Currently unused
+    # `metrics` must be either a list of callable metrics functions
+    if len(models) != len(data_generators):
+        if len(data_generators) != 1:
+            raise ValueError("Unequal number of models and data generators. Must be equal in numbers or use only a single data generator for all models.")
+        if len(data_generators) == 1:
+            data_generators = data_generators * len(models)
+    #
+    # Reformat metrics list to a metrics dictionary of the form
+    # {name: function pointer}.
+    if metrics is not None:
+        metrics_dict = {m.name: m for m in metrics}
+    #
+    model_count = 0
+    evaluations = list()
+    for m,g in zip(models, data_generators):
+        if metrics is None:
+            metrics_dict = {n.name: n for n in m.metrics}
+        #
+        # tf.data.Dataset family of generators in TensorFlow 2+
+        # Todo: could eventually be extended to include checks for 
+        # `dataset._dataset._batch_size` that is used in TensorFlow 1.*
+        if isinstance(g, dataset_ops.DatasetV2):
+            if hasattr(g, '_batch_size'):
+                if g._batch_size.numpy() != 1:
+                    raise ValueError("Batch size must be set to 1")
+        elif isinstance(g, TensorGenerator):
+            files = np.array([p.paths for p in g.path_iters]).flatten()
+            print(files)
+            g = DataGenerator(files, g)
+        else: # yolo
+            pass
```

Comment:
lol, at least log a warning

Reply:
This should throw an exception and captured by Keras/Tensorflow later if the generator is not legal